hono/src/services/items.ts

```diff
@@ -23,8 +23,38 @@ const selectItemSchema = createSelectSchema(items, {
   id: z.string().refine((val) => isCuid(val)),
 })
 
-app.get('/', injectDB, async (c) => {
-  return c.json(await c.get('db').query.items.findMany())
+const paginationValidation = z.object({
+  page: z
+    .string()
+    .refine((val) => !isNaN(+val) && +val >= 0, {
+      message: 'Page must be a non-negative number',
+    })
+    .optional(),
+  size: z
+    .string()
+    .refine((val) => !isNaN(+val) && +val > 0, {
+      message: 'Size must be a positive number',
+    })
+    .optional(),
+})
+
+app.get('/', zValidator('query', paginationValidation), injectDB, async (c) => {
+  const { page, size } = c.req.query()
+
+  const pageNumber: number = page ? +page : 0
+  const pageSize: number = size ? +size : 25
+
+  const p1 = c
+    .get('db')
+    .select()
+    .from(items)
+    // .where(eq(items.authorUsername, 'rak3rman')) TODO: replace with actual authorUsername
+    .prepare('p1')
+
+  return c.json({
+    items: (await p1.execute()).slice(pageNumber, pageNumber + pageSize),
+    total: (await p1.execute()).length,
```

Comment:
> There should be only one p1.execute() call, don’t review yet

*Just saw your comment above after I wrote up this response. Posting it anyway because it has some good things to consider.*

I don't think this is optimal. Consider the case where someone has 1000s of items. Then `p1.execute()` will return an object containing all 1000s of items. Now a MB size object is stored in RAM (not good) when we don't need it to be there. This concern exists for line 55 and 56.

The solution here is to leverage the utils provided by Postgres and Drizzle. Line 55: https://orm.drizzle.team/docs/select#limit--offset Line 56: could be solved by a simple `count(*)` sql statement which provides an exact count (which will be dead on arrival anyway considering concurrent writes), but there are other ways to return an "estimate" from the db which is good enough in our use case. Do some research and pick the cleanest implementation that makes sense for Shafa (I'd like to see something better than count(*) but not a 10 line Postgres wizard implementation).

Reply:
Agreed. Not sure if you saw this but I had this prior implementation [here](https://github.com/radicalplatforms/shafa/pull/110#discussion_r1513615014), which is technically a direct derivation of `count(*)` so I assume there is a better implementation. I'll see if there are better ways.

Reply:
No longer relevant, closing.